libs/libskynet/src/registry.ts

```diff
@@ -0,0 +1,200 @@
+import { blake2b } from "./blake2b.js";
+import { bufToB64, b64ToBuf, encodePrefixedBytes, encodeU64 } from "./encoding.js";
+import { addContextToErr } from "./err.js";
+import { Ed25519Keypair, ed25519KeypairFromEntropy, ed25519Sign, ed25519Verify } from "./ed25519.js";
+import { SEED_BYTES } from "./seed.js";
+import { sha512 } from "./sha512.js";
+import { Err } from "./types.js";
+
+// computeRegistrySignature will take a secret key and the required fields of a
+// registry entry and use them to compute a registry signature, returning both
+// the signature and the encoded data for the registry entry.
+function computeRegistrySignature(
+  secretKey: Uint8Array,
+  dataKey: Uint8Array,
+  data: Uint8Array,
+  revision: bigint
+): [signature: Uint8Array, err: Err] {
+  // Check that the data is the right size.
+  if (data.length > 86) {
+    return [new Uint8Array(0), "registry data must be at most 86 bytes"];
+  }
+
+  // Build the encoded data.
+  const [encodedData, errEPB] = encodePrefixedBytes(data);
+  if (errEPB !== null) {
+    return [new Uint8Array(0), addContextToErr(errEPB, "unable to encode provided registry data")];
+  }
+  const [encodedRevision, errEU64] = encodeU64(revision);
+  if (errEU64 !== null) {
+    return [new Uint8Array(0), addContextToErr(errEU64, "unable to encode the revision number")];
+  }
+
+  // Build the signing data.
+  const dataToSign = new Uint8Array(32 + 8 + data.length + 8);
+  dataToSign.set(dataKey, 0);
+  dataToSign.set(encodedData, 32);
+  dataToSign.set(encodedRevision, 32 + 8 + data.length);
+  const sigHash = blake2b(dataToSign);
+
+  // Sign the data.
+  const [sig, errS] = ed25519Sign(sigHash, secretKey);
+  if (errS !== null) {
+    return [new Uint8Array(0), addContextToErr(errS, "unable to sign registry entry")];
+  }
+  return [sig, null];
+}
+
+// deriveRegistryEntryID derives a registry entry ID from a provided pubkey and
+// datakey.
+function deriveRegistryEntryID(pubkey: Uint8Array, datakey: Uint8Array): [Uint8Array, string | null] {
+  // Check the lengths of the inputs.
+  if (pubkey.length !== 32) {
+    return [new Uint8Array(0), "pubkey is invalid, length is wrong"];
+  }
+  if (datakey.length !== 32) {
+    return [new Uint8Array(0), "datakey is not a valid hash, length is wrong"];
+  }
+
+  // Establish the encoding. First 16 bytes is a specifier, second 8
+  // bytes declares the length of the pubkey, the next 32 bytes is the
+  // pubkey and the final 32 bytes is the datakey. This encoding is
+  // determined by the Sia protocol.
+  const encoding = new Uint8Array(16 + 8 + 32 + 32);
+  // Set the specifier.
+  encoding[0] = "e".charCodeAt(0);
+  encoding[1] = "d".charCodeAt(0);
+  encoding[2] = "2".charCodeAt(0);
+  encoding[3] = "5".charCodeAt(0);
+  encoding[4] = "5".charCodeAt(0);
+  encoding[5] = "1".charCodeAt(0);
+  encoding[6] = "9".charCodeAt(0);
```

Comment:
How about:

```suggestion
  const algorithm = Array.from("ed25519").map(char => char.charCodeAt(0)); // encode algorithm as uint8
  encoding.set(algorithm, 0);
```
Easier to read the algorithm being used.